Common/Thread/ParallelLoop.cpp

```diff
@@ -0,0 +1,142 @@
+		for (int i = 0; i < range; i++) {
+			threadMan->EnqueueTaskOnThread(i, new LoopRangeTask(waitableCounter, loop, i, i + 1), TaskType::CPU_COMPUTE);
+		}
+		return waitableCounter;
+	} else {
+		// Split the range between threads. Allow for some fractional bits.
+		const int fractionalBits = 8;
+
+		int64_t totalFrac = (int64_t)range << fractionalBits;
+		int64_t delta = totalFrac / numTasks;
+
+		delta = std::max(delta, (int64_t)minSize << fractionalBits);
+
+		// Now we can compute the actual number of tasks.
+		// Remember that stragglers are done on the current thread
+		// so we don't round up.
+		numTasks = (int)(totalFrac / delta);
+		printf("numTasks: %d\n", numTasks);
+
+		WaitableCounter *waitableCounter = new WaitableCounter(numTasks);
+		int64_t counter = (int64_t)lower << fractionalBits;
+
+		// Split up tasks as equitable as possible.
+		for (int i = 0; i < numTasks; i++) {
+			int start = (int)(counter >> fractionalBits);
+			int end = (int)((counter + delta) >> fractionalBits);
+			if (end > upper) {
+				// Let's do the stragglers on the current thread.
+				break;
+			}
+			threadMan->EnqueueTaskOnThread(i, new LoopRangeTask(waitableCounter, loop, start, end), TaskType::CPU_COMPUTE);
+			counter += delta;
+			if ((counter >> fractionalBits) > upper) {
+				break;
+			}
+		}
+
+		// Run stragglers on the calling thread directly.
+		// We might add a flag later to avoid this for some cases.
+		int stragglerStart = (int)(counter >> fractionalBits);
+		int stragglerEnd = upper;
+		if (stragglerStart < stragglerEnd) {
+			// printf("doing stragglers: %d-%d\n", start, upper);
+			loop(stragglerStart, stragglerEnd);
+		}
+		return waitableCounter;
+	}
+}
+
+void ParallelRangeLoop(ThreadManager *threadMan, const std::function<void(int, int)> &loop, int lower, int upper, int minSize) {
+	if (cpu_info.num_cores == 1 || (minSize >= (upper - lower) && upper > lower)) {
+		// "Optimization" for single-core devices, or minSize larger than the range.
+		// No point in adding threading overhead, let's just do it inline (since this is the blocking variant).
+		loop(lower, upper);
+		return;
+	}
+
+	if (minSize < 1) {
+		// There's no obvious value to default to.
+		minSize = 1;
+	}
+
+	WaitableCounter *counter = ParallelRangeLoopWaitable(threadMan, loop, lower, upper, minSize);
+	// TODO: Optimize using minSize. We'll just compute whether there's a remainer, remove it from the call to ParallelRangeLoopWaitable,
+	// and process the remainder right here. If there's no remainer, we'll steal a whole chunk.
+	if (counter) {
+		counter->WaitAndRelease();
+	}
+}
+
+// NOTE: Supports a max of 2GB.
+void ParallelMemcpy(ThreadManager *threadMan, void *dst, const void *src, size_t bytes) {
+	// This threshold can probably be a lot bigger.
+	if (bytes < 512) {
+		memcpy(dst, src, bytes);
+		return;
+	}
+
+	// 128 is the largest cacheline size on common CPUs.
+	// Still I suspect that the optimal minSize is a lot higher.
```

Comment:
My tests had shown 128 * 1024 to be a much better threshold.  I didn't try exhaustively, but it was definitely better than 32 KB or 512 KB, iirc.  Obviously this varies by CPU, but I was assuming that preloading and pipelining in `memmove()` were a factor.

-[Unknown]

Reply:
Sounds good, changing.